### ISSUE TITLE
Build tool provenance/record and versions

### ISSUE THREAD

**Issue by natefoo** [refgenie/refgenie#195]
I was digging a bit to find out if the `rna_star` assets were built with STAR 2.7.x since I believe they're not compatible with older versions, and that raised a couple of questions for me:

1. Would it be possible to record a few things, such as the identifiers of the containers used to build, and the versions of the commands? For example, the [build recipe for hg38/star_index](http://awspds.refgenie.databio.org/hg38/build_recipe_star_index__default.json) shows the build container as `databio/refgenie` (a bulker image?) but the digest of the container(s) used isn't recorded in the build log. Additionally, the version of STAR isn't captured. If the recipe included a method for fetching this (e.g. `STAR --version` or whatever), I think that'd be helpful.
2. Any thoughts on versioning the asset names or indicating their compatible software versions? If STAR index formats change in 2.8.x, for example.

**Comment by nsheff** [comment]
Yes, both of these are things that we have been thinking about for awhile. 

TL; DR: Both of these are things that we are planning to accommodate when we move to allowing custom recipes. 

1. Yes, you are right that we need this. Right now, the recipe format doesn't accommodate it, but when we revamp that format, we had planned to add a way for the recipe author to record custom key-value pairs. The idea is that the recipe would provide something like:

```
custom_properties:
  version: `STAR --version`
  method: "xyz"
```

The commands encoded in this way in the recipe would be run in the same environment as the build command, and we'd record the custom outputs. these would then be accessible via API, and they'd automatically update when the recipe was built with a different container, for example. I think along with this we should also automatically keep the hash of the actual container that was used. the `databio/refgenie` is actually a container, not a bulker manifest, as that's how we were originally building them, but you're right that this is only specifying what the recipe suggests, not what was actually run...

2. yeah, I've been thinking about this some. the way I was going to approach this was using tags. Right now we've simplified things by just using the 'default' tag for everything. But now that we're starting to deal with versions like this, it's probably time for the server to actually specify tags.  So, the way I think we should do this is that we should determine the tag that we want to use for the asset and then build and maintain multiple assets that way. So we'd tag star index with 2.7

You'd say: `refgenie pull hg38/star_index:2.7` or whatever. We could even make the tags accept the custom properties above, so you'd do something like:

```
default_tag: "{version}"
```

When this recipe gets created, it would automatically tag it with the version, as specified by the software.

**Comment by nsheff** [comment]
Related to #170 

An easy short-term solution is that the build logs should at least show the versions. For some tools this happens automatically... but STAR doesn't output its version by default, so it's not in the log:

http://refgenomes.databio.org/v2/asset/hg38/star_index/splash?tag=default

so we should at the very least add `STAR --version` as a command in that recipe, which we could already do even without the recipe format revamp described above.

**Comment by bgruening** [comment]
I like this a lot. We need to have some way of recording the version explicitly. Often it will be just a data of download and assembly. I would propose to have `tag: "{version}"` as a mendatory field, but have the key-value pair idea for logging and provenance.

**Comment by pvanheus** [comment]
In the medium term, if refgenie build tools were CommonWL tools or workflows, `cwltool --provenance` could generate a provenance record for each run. At the moment this makes a "fat" research object with the input and output data included but a "slim" mode could be made just storing info about inputs and outputs and runs. I feel like going down this path gets a lot of tool integration "for free".

**Comment by nsheff** [comment]
@pvanheus Yes! fact, I have been hoping to simply use CWL directly for the recipes. I hope it's possible. I was imagining that the recipe format would include a CWL workflow as the actual recipe, and then just add to it whatever metadata we need... I'm hoping to dive into this deeply now and in the next week or two I'll try to post a proposal for a recipe structure, which I'd love to get your feedback on.
